packages/auth/src/providers/cognito/utils/signInHelpers.ts

```
@@ -648,3 +657,68 @@ export async function assertUserNotAuthenticated() {
 		});
 	}
 }
+
+/**
+ * This function is used to kick off the device management flow.
+ *
+ * If an error is thrown while generating a hash device or calling the `ConfirmDevice`
+ * client, then this API will ignore the error and return undefined. Otherwise the authentication
+ * flow will not complete and the user won't be able to be signed in.
+ *
+ * @returns DeviceMetadata | undefined
+ */
+export async function getNewDeviceMetatada(
+	amplify: AmplifyClassV6,
+	newDeviceMetadata?: NewDeviceMetadataType,
+	accessToken?: string
+): Promise<DeviceMetadata | undefined> {
+	const config = amplify.getConfig().Auth?.Cognito;
+	assertTokenProviderConfig(config);
+	const userPoolName = config.userPoolId.split('_')[1] || '';
+	const authenticationHelper = new AuthenticationHelper(userPoolName);
+	const deviceKey = newDeviceMetadata?.DeviceKey;
+	const deviceGroupKey = newDeviceMetadata?.DeviceGroupKey;
+
+	return new Promise((resolve, _) => {
+		authenticationHelper.generateHashDevice(
+			deviceGroupKey ?? '',
+			deviceKey ?? '',
+			async (errGenHash: unknown) => {
+				if (errGenHash) {
+					// TODO: log error here
+					resolve(undefined);
+					return;
+				}
+
+				const deviceSecretVerifierConfig = {
+					Salt: toBase64(fromHex(authenticationHelper.getSaltToHashDevices())),
+					PasswordVerifier: toBase64(
+						fromHex(authenticationHelper.getVerifierDevices())
+					),
+				};
+
+				const randomPassword = authenticationHelper.getRandomPassword();
+
+				try {
+					await confirmDevice(
+						{ region: getRegion(config.userPoolId) },
+						{
+							AccessToken: accessToken,
+							DeviceKey: newDeviceMetadata?.DeviceKey,
+							DeviceSecretVerifierConfig: deviceSecretVerifierConfig,
+						}
+					);
+
+					resolve({
+						deviceKey,
+						deviceGroupKey,
+						randomPassword,
+					});
+				} catch (error) {
+					// TODO: log error here
+					resolve(undefined);
```

why not reject?

---

because otherwise the user will not be authenticated.

Here's a closed https://github.com/aws-amplify/amplify-js/issues/2087 that still is relevant . Confirmed that swift does the same. They swallowed the error and authenticate the user.